src/scheduler/JobScheduler.ts

```diff
@@ -149,23 +151,57 @@ export class JobScheduler {
           : jobEntity.concurrency;
       this.logger.debug('execute job', { name: this.jobName, times: numToExecute });
 
-      for (let i = 0; i < numToExecute; i++) {
+      this.executorError = new Array(numToExecute);
+
+      for (let instanceNumber = 0; instanceNumber < numToExecute; instanceNumber++) {
+        this.logger.debug('executing job instance', { instanceNumber });
+        this.executorError[instanceNumber] = false;
+
+        const executionTimeout =
+          jobEntity.timeout !== undefined
+            ? setSafeTimeout(
+                async () => this.handleTimeoutReached(instanceNumber),
+                jobEntity.timeout,
+                this.logger,
+                'error occurred in timeout',
+              )
+            : undefined;
+
         // eslint-disable-next-line no-void
-        void this.jobExecutor.execute(jobEntity, parameters).catch((e) => {
-          this.handleUnexpectedError(e);
-        });
+        void this.jobExecutor
+          .execute(jobEntity, parameters)
+          .catch((e) => {
+            this.handleUnexpectedError(instanceNumber, e);
+          })
+          .finally(() => {
+            if (this.executorError[instanceNumber] !== undefined) {
+              this.logger.debug('job executed correctly, cancel timout', { instanceNumber });
+              clearTimeout(executionTimeout);
+            }
+          });
       }
     } catch (e) {
-      this.handleUnexpectedError(e);
+      this.handleUnexpectedError(-1, e);
     }
   }
 
-  private handleUnexpectedError(error: unknown): void {
+  private handleTimeoutReached(instanceNumber: number): void {
+    this.logger.error('timeout reached', MomoErrorType.executeJob, { name: this.jobName, instanceNumber });
+    // TODO how to clean up? what to do now?
+    // clean up only the failed execution? But we do not store every execution of a concurrent job separately, we only have one executionInfo. Change it to an array of length maxRunning?
+    // or clean up everything - that is, stop the job and restart it?
+  }
```

Comment:
Changing the execution of a concurrent job to an array would create non-trivial backwards compatibility issues. And I'm not 100% certain it's necessary. 

I guess we could somehow mark the job itself as "failed" and just let the scheduler try to run it again next time it it's time comes? Or perhaps we could force it to wait `2*timeout`, so we know any currently running job instance is already finished. 

This is a case that should happen very rarely anyway, so I don't think simply updating the repo to say "this job failed" would cause too many issues.

Reply:
We had a talk with @set3mu and were wondering if it might make sense to just decrement the number of current executions by 1 and set the "lastFinished" and "lastResult" to now and "failed" (or "timedOut") respectively. That should ensure the next time the job is meant to run it gets to, right?

The only question is what to do if we get an error when decrementing the number of current executions. 🤔 Then we're right back to square one.

Reply:
But do we know that "just decrement the number of current executions by 1" will always be the fix? What if the "unexpected error" happens not on decreasing but increasing the "running" field?
I'm not really sure, but probably this could happen:
momo wants to execute job
momo starts the timeout
db connection fails when trying to increasing the "running" field, error is thrown and thus timeout is not cleared (would be cleared in line 179, but because of the error we end up in line 184)
after timeout we say "ok, it's dead, let's decrease "running" by 1"

and that is just not correct.
So I guess it would be safer to basically throw away everthing we now about the job and start over from scratch.

Reply:
But what happens when we stop and restart a job? The current implementation of JobScheduler/JobExecutor will not stop running instances of the job if it is stopped. So I tell it to stop, but the currently running instances will continue to run in the background.
If I restart the job and it triggers more instances of the job, we might end up running more instances of the job in parallel that intended. If users of momo really rely on the "maxRunning" that might be a problem for them.

Can we change the JobScheduler to actually cancel running instances when we stop it? Can cancelling running jobs cause problems for other users (probably not every job is happily cancelable at any given point in time ...)?

Reply:
What we could do - stop the job, want for <timeout> and only then restart. Then we give responsibility to the user (who defines the timeout) to ensure that we do not exceed the maxRunning ... However users maybe want to make the timeout rather big. Which would mean we wait rather long before restarting the job - we basically wait 2*timeout before we restart the broken job = one timeout until we consider it dead + one timeout before we consider also all other potentially running instances as finished and are safe to restart the job.



However, in case maxRunning is 1 (or 0, meaning infinity) it would be safe to just restart a broken job immediately.